File: mozn-dashboard/web/types/thresholds.ts
/**
 * Thresholds API contract types. Mirror the JSON served by the Go backend
 * (internal/model/thresholds.go). Part of the shared types/ contract layer so
 * the fetch layer (lib/api) can reference them without reaching into features/.
 * The thresholds feature re-exports these from features/thresholds/types.
 */
import type { ThresholdMetric } from "@/types/shared";

export type { ThresholdMetric };

export type TierMode = "auto" | "manual";

export interface ThresholdTier {
  /** Backend AlertThreshold row id — targets PUT/DELETE for this tier. */
  id: string;
  /** Raw backend identity (immutable on PUT): parameter + severity + applies_to. */
  parameter: string;
  severity: "yellow" | "orange" | "red";
  appliesTo: "observed" | "forecast" | "both";
  isActive: boolean;
  sustainDurationMinutes: number | null;
  name: string;
  mode: TierMode;
  description: string;
  value: string;
  unit: string;
  sustained: string;
}

export interface ScaleStop {
  label: string;
  value: string;
  /** Token name: normal | advisory | watch | warning. */
  tone: string;
}

export interface MetricThresholds {
  /** Backend parameter key — the card identity. One card per parameter, so
   *  temp_high_c and temp_low_c (etc.) no longer collapse into one card and
   *  hide the second parameter's tiers. */
  parameter: string;
  /** English display label (from paramLabel); Arabic applied via td() in the UI. */
  label: string;
  unit: string;
  /** Below-comparison parameter (temp_low_c): breaches when value < cut-off, so
   *  the editor renders ≤ and inverts the tier ordering. Mirrors isValueBreached
   *  in the Go engine (services/alert_checker.go). */
  isLowSide: boolean;
  perStationOverrides: boolean;
  /** Region the thresholds belong to — used for the live impact preview. */
  regionId: string;
  /** Region's English name (translated in the UI via region.* keys). */
  regionName: string;
  tiers: ThresholdTier[];
  scale: ScaleStop[];
}

export interface ThresholdChange {
  /** History row id — sent as {history_id} to the revert endpoint. */
  id: string;
  /** The threshold row the history belongs to — the revert URL target. */
  thresholdId: string;
  change: string;
  by: string;
  when: string;
  /** Raw ISO of the change — rendered client-side so it re-localizes instantly. */
  whenAt: string;
}

/** Comparator for a compound-rule condition (backend: gt/gte/lt/lte/eq). */
export type CompoundOperator = "gt" | "gte" | "lt" | "lte" | "eq";

/** One condition inside a compound rule (backend RuleCondition). */
export interface CompoundCondition {
  id?: string;
  parameter: string;
  operator: CompoundOperator;
  value: number;
  sustainMinutes: number | null;
}

/** A multi-metric compound rule (A3.0). All conditions must hold to fire. */
export interface CompoundRule {
  id: string;
  name: string;
  regionId: string;
  severity: "yellow" | "orange" | "red";
  isActive: boolean;
  conditions: CompoundCondition[];
}

export interface ThresholdsPage {
  metrics: MetricThresholds[];
  changes: ThresholdChange[];
}

File: mozn-dashboard/server/internal/model/thresholds.go
package model

// ThresholdMetric identifies a measured metric.
type ThresholdMetric string

const (
	MetricRainfall    ThresholdMetric = "rainfall"
	MetricWind        ThresholdMetric = "wind"
	MetricWater       ThresholdMetric = "water"
	MetricTemperature ThresholdMetric = "temperature"
)

// TierMode is whether a tier updates automatically or needs manual action.
type TierMode string

const (
	TierAuto   TierMode = "auto"
	TierManual TierMode = "manual"
)

// ThresholdTier is one severity tier within a metric.
type ThresholdTier struct {
	Name        string   `json:"name"`
	Mode        TierMode `json:"mode"`
	Description string   `json:"description"`
	Value       string   `json:"value"`
	Unit        string   `json:"unit"`
	Sustained   string   `json:"sustained"`
}

// ScaleStop is one labeled stop on the threshold scale bar.
type ScaleStop struct {
	Label string `json:"label"`
	Value string `json:"value"`
	Tone  string `json:"tone"`
}

// MetricThresholds groups the tiers and scale for one metric.
type MetricThresholds struct {
	Metric              ThresholdMetric `json:"metric"`
	Label               string          `json:"label"`
	Unit                string          `json:"unit"`
	PerStationOverrides bool            `json:"perStationOverrides"`
	Tiers               []ThresholdTier `json:"tiers"`
	Scale               []ScaleStop     `json:"scale"`
}

// ImpactStation is a station whose state would change if thresholds applied.
type ImpactStation struct {
	Name string `json:"name"`
	Tier string `json:"tier"`
	Tone string `json:"tone"`
	Note string `json:"note,omitempty"`
}

// ThresholdChange is one row of the threshold change history.
type ThresholdChange struct {
	ID     string `json:"id"`
	Change string `json:"change"`
	By     string `json:"by"`
	When   string `json:"when"`
}

// ThresholdImpact is the impact-preview block.
type ThresholdImpact struct {
	Note     string          `json:"note"`
	Stations []ImpactStation `json:"stations"`
}

// ThresholdsPage is the payload for the Alerts & Thresholds screen.
type ThresholdsPage struct {
	Metrics []MetricThresholds `json:"metrics"`
	Impact  ThresholdImpact    `json:"impact"`
	Changes []ThresholdChange  `json:"changes"`
}
